src/components/CodeMirrorEditor.jsx

```diff
@@ -0,0 +1,90 @@
+import CodeMirror from 'codemirror';
+import isNil from 'lodash-es/isNil';
+import LRU from 'lru-cache';
+import PropTypes from 'prop-types';
+import React, {useEffect, useRef} from 'react';
+
+import 'codemirror/mode/htmlmixed/htmlmixed';
+import 'codemirror/mode/css/css';
+import 'codemirror/mode/javascript/javascript';
+
+const CODEMIRROR_MODES_MAP = {
+  html: 'htmlmixed',
+  css: 'css',
+  javascript: 'javascript',
+};
+
+export default function CodeMirrorEditor({
+  language,
+  projectKey,
+  source,
+  onInput,
+}) {
+  const containerRef = useRef();
+  const editorRef = useRef();
+  const docsCacheRef = useRef(new LRU(3));
+  const mode = CODEMIRROR_MODES_MAP[language];
+
+  useEffect(() => {
+    const {current: container} = containerRef;
+
+    const editor = editorRef.current = CodeMirror(
+      container,
+      {
+        lineNumbers: true,
+        lineWrapping: true,
```

Comment:
You haven't provided a theme name or mode to the initialize call- I think you can, since the props are available here

Reply:
As for a theme name, as I mentioned, I have yet to land on a theme that is obviously preferable to the default theme.

For the mode, it would feel good to put it here, but that would make the `language` prop an effect dependency, which we definitely don't want here! Though practically we would never expect `language` to change, if it did in fact change, we should not tear down and re-create the editor instance; we should just swap in a new `Doc` instance, with the appropriate mode. And that's exactly what the following effect hook does.

Reply:
That wasn't obvious to me on first review, I'd appreciate a comment on why the mode is missing